azkaban-execserver/src/main/java/azkaban/execapp/event/JobCallbackRequestMaker.java

```diff
@@ -0,0 +1,235 @@
+  private static final int DEFAULT_RESPONSE_WAIT_TIME_OUT_MS = 5000;
+  private static final int MAX_RESPONSE_LINE_TO_PRINT = 50;
+
+  private static final int DEFAULT_THREAD_POOL_SIZE = 10;
+
+  private static JobCallbackRequestMaker instance;
+  private static boolean isInitialized = false;
+
+  private FutureRequestExecutionService futureRequestExecutionService;
+  private int responseWaitTimeoutMS = -1;
+
+  public static void initialize(Props props) {
+    if (props == null) {
+      throw new NullPointerException("props argument can't be null");
+    }
+    instance = new JobCallbackRequestMaker(props);
+    isInitialized = true;
+    logger.info("Initialization for " + JobCallbackRequestMaker.class.getName()
+        + " is completed");
+  }
+
+  public static boolean isInitialized() {
+    return isInitialized;
+  }
+
+  public static JobCallbackRequestMaker getInstance() {
+    if (!isInitialized) {
+      throw new IllegalStateException(JobCallbackRequestMaker.class.getName()
+          + " hasn't initialzied");
+    }
+    return instance;
+  }
+
+  private JobCallbackRequestMaker(Props props) {
+
+    int connectionRequestTimeout =
+        props.getInt("jobcallback.connection.request.timeout",
+            DEFAULT_TIME_OUT_MS);
+
+    int connectionTimeout =
+        props.getInt("jobcallback.connection.timeout", DEFAULT_TIME_OUT_MS);
+
+    int socketTimeout =
+        props.getInt("jobcallback.socket.timeout", DEFAULT_TIME_OUT_MS);
+
+    responseWaitTimeoutMS =
+        props.getInt("jobcallback.response.wait.timeout",
+            DEFAULT_RESPONSE_WAIT_TIME_OUT_MS);
+
+    logger.info("responseWaitTimeoutMS: " + responseWaitTimeoutMS);
+
+    RequestConfig requestConfig =
+        RequestConfig.custom()
+            .setConnectionRequestTimeout(connectionRequestTimeout)
+            .setConnectTimeout(connectionTimeout)
+            .setSocketTimeout(socketTimeout).build();
+
+    logger.info("Global request configuration " + requestConfig.toString());
+
+    HttpClient httpClient =
+        HttpClientBuilder.create().setDefaultRequestConfig(requestConfig)
+            .build();
+
+    int jobCallbackThreadPoolSize =
+        props.getInt("jobcallback.thread.pool.size", DEFAULT_THREAD_POOL_SIZE);
+    logger.info("Jobcall thread pool size: " + jobCallbackThreadPoolSize);
+
+    ExecutorService executorService =
+        Executors.newFixedThreadPool(jobCallbackThreadPoolSize);
+    futureRequestExecutionService =
+        new FutureRequestExecutionService(httpClient, executorService);
+  }
+
+  public FutureRequestExecutionMetrics getJobcallbackMetrics() {
+    return futureRequestExecutionService.metrics();
+  }
+
+  public void makeHttpRequest(String jobId, Logger logger,
+      List<HttpRequestBase> httpRequestList) {
+
+    if (httpRequestList == null || httpRequestList.isEmpty()) {
+      logger.info("No HTTP requests to make");
+      return;
+    }
+
+    for (HttpRequestBase httpRequest : httpRequestList) {
```

Comment:
Here, the code is doing all the http call sequentially: one task, wait for the task to complete, then next http call. It is not quite efficient. Maybe should submit all requests to future service, then check the status of all the future tasks?